mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp

```diff
@@ -398,8 +399,15 @@ LogicalResult GPUPrintfOpToHIPLowering::matchAndRewrite(
   mlir::Type llvmI64 = typeConverter->convertType(rewriter.getI64Type());
   // Note: this is the GPUModule op, not the ModuleOp that surrounds it
   // This ensures that global constants and declarations are placed within
-  // the device code, not the host code
-  auto moduleOp = gpuPrintfOp->getParentOfType<gpu::GPUModuleOp>();
+  // the device code, not the host code.
+  Operation *moduleOp = gpuPrintfOp->getParentOfType<gpu::GPUModuleOp>();
+  // However, if the `gpu.module` is already lowered or for compilers that don't
+  // use `gpu.module`, fall back to `builtin.module`.
+  if (!moduleOp)
+    moduleOp = gpuPrintfOp->getParentOfType<ModuleOp>();
```

Comment:
Please look for the closest SymbolTable parent instead instead of hardcoding specific ops.

Reply:
```
Operation *symtableOp = op->getParentWithTrait<OpTrait::SymbolTable>();
```

There is also:

```
Operation *symtableOp = SymbolTable::getNearestSymbolTable(op);
```

Which has a subtle difference with unregistered ops...

Reply:
Thanks, updated!